knimin/lib/data_access.py

```diff
@@ -4258,6 +4284,84 @@ def delete_shotgun_plate(self, plate_id):
                 # plate_id did not exist
                 raise ValueError('Shotgun Plate %s does not exist' % plate_id)
 
+    def quantify_shotgun_plate(self, shotgun_plate_id, email, volume,
+                               plate_reader, plate_concentration):
+        """Adds the DNA quantification information to the shotgun plate
+
+        Parameters
+        ----------
+        shotgun_plate_id : int
+            The shotgun plate id
+        email : str
+            The email of the user
+        volume : float
+            The volume used for DNA quantification
+        plate_reader : str
+            The plate reader used
+        plate_concentration : 2d numpy array of floats
+            The per-well DNA concentration
+
+        Raises
+        ------
+        ValueError
+            If `plate_concentration` dimensions doesn't match the plate type
+        """
+        with TRN:
+            sgp = self.read_shotgun_plate(shotgun_plate_id)
+            rp = dict(self.read_plate_type(sgp['plate_type_id']))
+            pc_rows, pc_cols = plate_concentration.shape
+            processing_robot_id = self.get_or_create_property_option_id(
+                "plate_reader", plate_reader)
+
+            if pc_rows != rp['rows'] or pc_cols != rp['cols']:
+                raise ValueError('plate_concentration wrong shape, should '
+                                 'be: (%d, %d) but is: (%d, %d)' % (
+                                    rp['rows'], rp['cols'], pc_rows, pc_cols))
+            # we expect 4 plates, thus 4 empty rows
+            orders = [[], [], [], []]
+            for pid, order in sgp['condensed_plates']:
+                sp_id = self.read_dna_plate(pid)['sample_plate_id']
+                orders[order] = self.read_sample_plate_layout(sp_id)
+
+            # now we start creating the merged plate
+            data = []
+            for rid, row in enumerate(zip_longest(*orders, fillvalue=None)):
+                for cid, (a, b, c, d) in enumerate(zip_longest(*row,
+                                                   fillvalue=None)):
+                    new_cid = 2 * cid
+                    new_rid = 2 * rid
+                    data.append([
```

Comment:
can this be decomposed into a local method? eg

```python
def set_grid(sample_id, plate_id, concentration, row_offset, col_offset, block_row_offset, block_col_offset, data):
    grid_row = row_offset + block_row_offset
    grid_col = col_offset + block_col_offset
    data.append((plate_id, grid_row, grid_col, sample_id, concentration[grid_row, grid_col]))
```

Then the following lines are:

```python
# set the pattern:
# a b
# c d
set_grid(a['sample_id'], shotgun_plate_id, new_rid, new_cid, 0, 0, data)
set_grid(b['sample_id'], shotgun_plate_id, new_rid, new_cid, 0, 1, data)
set_grid(c['sample_id'], shotgun_plate_id, new_rid, new_cid, 1, 0, data)
set_grid(d['sample_id'], shotgun_plate_id, new_rid, new_cid, 1, 1, data)
```